### PR TITLE
Cell annotations to factors

### PR DESCRIPTION
**Pull request by keller-mark**
This is basically a duplicate of https://github.com/hubmapconsortium/portal-containers/pull/32 except here the source data file is hard-coded and comes from the vitessce-data S3.

### CODE REVIEW

File: process.sh [hunk 1]
```diff
@@ -15,7 +15,7 @@ main() {
     INPUT="$FILES/input"
     OUTPUT="$FILES/output"
 
-    for DATASET in cytokit linnarsson dries wang cao spraggins; do
+    for DATASET in cytokit linnarsson dries wang cao spraggins annotations_spleen_0510; do
```
Comment (mccalluc) [inline process.sh:18]:
```suggestion
    for DATASET in cytokit linnarsson dries wang cao spraggins satija; do
```
Just a convention, but we've been generally naming by PI.

File: python/annotations_csv_reader.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+class COLUMNS(Enum):
+    # Cell Type Annotation column names
+    CELL_ID = "cell_id"
+    ANNOTATION = "annotation"
+    PREDICTION_SCORE = "prediction_score"
+    DATASET_ID = "dataset_id"
+
+
+def factors_json(df, dataset_id):
```
Comment (mccalluc) [inline python/annotations_csv_reader.py:17]: This is picky, but this doesn't really have anything to do with JSON, which is a serialization format. It returns a dict.

It also doesn't really tell us what it does. Verbs are good. Maybe `grep_dataframe_by_id`?

File: python/annotations_csv_reader.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    annotation_values = df[COLUMNS.ANNOTATION.value].unique().tolist()
+
+    # Append a temporary column to the dataframe containing
+    # the index of each annotation,
+    # relative to the `annotation_values` list.
+    df["annotation_index"] = df[COLUMNS.ANNOTATION.value].apply(
+        lambda val: annotation_values.index(val)
+    )
+
+    # Construct the factors dict, which will be saved as a JSON object.
```
Comment (mccalluc) [inline python/annotations_csv_reader.py]:
```suggestion
    # Construct the factors dict.
```

File: python/annotations_csv_reader.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        'metadata from CSV.')
+    parser.add_argument(
+        '--csv_file', required=True,
+        help='CSV file containing annotations.')
+    parser.add_argument(
+        '--dataset_id', type=str,
+        help='Filter by this dataset ID.')
+    parser.add_argument(
+        '--factors_file', type=argparse.FileType('w'),
+        help='Write the cell factors to this file.')
```
Comment (mccalluc) [inline python/annotations_csv_reader.py:86]: Should these be required too?

File: python/annotations_csv_reader.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    with open(args.csv_file) as csv_file:
+        df = pd.read_csv(csv_file)
+
+    if args.factors_file:
+        json.dump(
+            factors_json(df, args.dataset_id),
+            args.factors_file,
+            indent=1
+        )
```
Comment (mccalluc) [inline python/annotations_csv_reader.py:97]:
```suggestion
    json.dump(
        factors_json(df, args.dataset_id),
        args.factors_file,
        indent=1
    )
```

File: python/annotations_csv_reader.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    # Append a temporary column to the dataframe containing
+    # the index of each annotation,
+    # relative to the `annotation_values` list.
+    df["annotation_index"] = df[COLUMNS.ANNOTATION.value].apply(
+        lambda val: annotation_values.index(val)
+    )
+
+    # Construct the factors dict, which will be saved as a JSON object.
+    factors = {
+        "Cell Type Annotations": {
```
Comment (mccalluc) [inline python/annotations_csv_reader.py:63]: - This is arbitrary, right? Consider exposing at a higher level to make change easier... but you don't want to make the interface too complicated, either. Use your judgment.
- Reference the schema which dictates this particular form.

File: requirements.txt [hunk 7]
```diff
@@ -9,4 +9,5 @@ pyimzml==1.3.0
 dask==2.11.0
 toolz==0.9.0
 fsspec==0.3.3
-tifffile==2020.2.16
\ No newline at end of file
+tifffile==2020.2.16
+scikit-image==0.16.1
```
Comment (mccalluc) [inline requirements.txt:13]:
```suggestion
```

I don't see this being used? Sorry if I'm missing something obvious.

File: scripts/process_annotations_spleen_0510.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    get_CLI_args "$@"
+
+    DATA_TITLE='annotations_spleen_0510'
+    CSV_IN="$INPUT/annotations_spleen_0510.csv"
+
+    echo "Download and process cells..."
+
+    [ -e "$CSV_IN" ] || \
+        wget "https://vitessce-data.s3.amazonaws.com/source-data/annotations_spleen_0510/annotations_spleen_0510.csv" -O "$CSV_IN"
```
Comment (mccalluc) [inline scripts/process_annotations_spleen_0510.sh:20]: Explain where this comes from, or better, is there a URL we can use which NYGC maintains?

File: scripts/process_annotations_spleen_0510.sh [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    echo "Running: $CMD"
+    eval $CMD
+
+    CMD="$BASE/python/annotations_csv_reader.py --csv_file $CSV_IN --factors_file $OUTPUT/HBM984.GRBB.858.factors.json --dataset_id HBM984.GRBB.858"
+    echo "Running: $CMD"
+    eval $CMD
+
+    CMD="$BASE/python/annotations_csv_reader.py --csv_file $CSV_IN --factors_file $OUTPUT/HBM336.FWTN.636.factors.json --dataset_id HBM336.FWTN.636"
+    echo "Running: $CMD"
+    eval $CMD
```
Comment (mccalluc) [inline scripts/process_annotations_spleen_0510.sh:42]:
```suggestion
    for ID in HBM472.NTNN.543 HBM556.QMSM.776 HBM396.RPRR.624 HBM984.GRBB.858 HBM336.FWTN.636; do
        CMD="$BASE/python/annotations_csv_reader.py
            --csv_file $CSV_IN
            --factors_file $OUTPUT/$ID.factors.json
            --dataset_id $ID"
        echo "Running: $CMD"
        eval $CMD
    done
```

Loop!?

**Review by mccalluc — COMMENTED** [review]
Lots of little comments, nothing big. Address as you think best, and merge. Thanks!

File: scripts/process_annotations_spleen_0510.sh (hunk 8 above)
Comment (keller-mark) [reply scripts/process_annotations_spleen_0510.sh:20]: I got this from slack, which Nils got from an email

File: requirements.txt (hunk 7 above)
Comment (keller-mark) [reply requirements.txt:13]: Ilan asked the same thing in the other PR: https://github.com/hubmapconsortium/vitessce-data/pull/94#discussion_r426083253

**Comment by keller-mark** [comment]
Actually going to close this and the corresponding PR in `portal-containers` since they are pretty much duplicates and could both be improved. Will replace with a new PR shortly...

**Pull request closed without merging**